src/server.js

```diff
@@ -0,0 +1,276 @@
+/**
+ * KumaScript HTTP service
+ *
+ * This file defines the Server class, an Express-based HTTP server
+ * for rendering KumaScript macros within HTML documents.
+ *
+ * The endpoint for rendering documents is a POST to /doc/, with
+ * environment variables encoded in request headers.
+ *
+ * The server also supports these endpoints:
+ *
+ * - GET /macros/ returns a list of defined macros
+ * - GET /revision/ returns the git revision of this version of KumaScript
+ * - GET /healthz/ returns a 204 status if the server is running
+ * - GET /readiness/ returns 204 if the server is ready for requests
+ *
+ * @prettier
+ */
+const path = require('path');
+const url = require('url');
+
+const express = require('express');
+const morgan = require('morgan');
+const request = require('request');
+const winston = require('winston');
+
+const config = require('./config.js');
+const firelogger = require('./firelogger.js');
+const Templates = require('./templates.js');
+const render = require('./render.js');
+
+// TODO: figure out the newrelic config details and re-enable before deploying
+//var newrelic_conf = ks_conf.nconf.get('newrelic');
+//if (newrelic_conf && newrelic_conf.license_key) require('newrelic');
+
+class Server {
+    /**
+     * Build the service, but do not listen yet.
+     */
+    constructor() {
+        const app = (this.app = express());
+
+        // This will raise an error if the macros are not found
+        this.templates = new Templates(config.macrosDirectory);
+
+        // Configure a logger that pipes to the winston logger.
+        if (config.logging) {
+            var logger = morgan(
+                ':remote-addr - - [:date] ":method :url ' +
+                    'HTTP/:http-version" :status :res[content-length] ' +
+                    '":referrer" ":user-agent" :response-time',
+                {
```

Comment:
I think we should toss this second argument to the `morgan` constructor, so we'd just end-up using the default stream which is `process.stdout`. The second argument messes up the logging, in that we get the object as a string, like this:
```
{"source":"server","pid":1,"level":"info","message":"::ffff:192.168.80.1 - - [Tue, 15 Jan 2019 20:00:16 GMT] \"GET /macros HTTP/1.1\" 304 - \"-\" \"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.14; rv:64.0) Gecko/20100101 Firefox/64.0\" 102.841"}
```
rather than, without it, the formatted string, like this:
```
::ffff:192.168.48.1 - - [Tue, 15 Jan 2019 18:24:56 GMT] "GET /macros HTTP/1.1" 304 - "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.14; rv:64.0) Gecko/20100101 Firefox/64.0" 3.196
```

That would get us one step closer to dumping `winston` as well.

Reply:
Fixed in af2b87f.

Thanks for pointing this out. I was cargo-culting the old logging code. The logs are much cleaner now!

Reply:
LOL, love that phrase "cargo-culting", which I hadn't heard before! 😆